File: Data Structures & Algorithms/daily-temperatures/submission-1.java
// Time Complexity: O(n), each element pushed and popped from stack once
// Space Complexity: O(n), stack can store up to n elements in worst case

class Solution {
    public int[] dailyTemperatures(int[] temperatures) {
        int[] res = new int[temperatures.length];
        // int[] contains [value, index]
        Stack<int[]> stack = new Stack<>();

        for (int i = temperatures.length - 1; i >= 0; i--) {
            int currTemp = temperatures[i];

            while (!stack.isEmpty() && stack.peek()[0] <= currTemp) {
                stack.pop();
            }

            if (stack.isEmpty()) {
                res[i] = 0;
            } else {
                res[i] = stack.peek()[1] - i;
            }

            stack.push(new int[]{currTemp, i});
        }

        return res;
    }
}

File: Data Structures & Algorithms/minimum-stack/submission-1.py
# Overall Space Complexity: O(n), both stacks can store up to n elements
class MinStack:
    def __init__(self):      
        self.stack = []
        self.minStack = []  

    # Time Complexity: O(1)
    def push(self, val: int) -> None:
        self.stack.append(val)
        if len(self.minStack) == 0 or self.minStack[len(self.minStack) - 1] >= val:
            self.minStack.append(val)

    # Time Complexity: O(1)
    def pop(self) -> None:
        poppedVal = self.stack.pop()
        if self.minStack[len(self.minStack) - 1] == poppedVal:
            self.minStack.pop()
        
    # Time Complexity: O(1)
    def top(self) -> int:
        return self.stack[len(self.stack) - 1]

    # Time Complexity: O(1)
    def getMin(self) -> int:
        return self.minStack[len(self.minStack) - 1]
